packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebViewClient.java

```diff
@@ -0,0 +1,112 @@
+package io.flutter.plugins.webviewflutter;
+
+import android.annotation.TargetApi;
+import android.graphics.Bitmap;
+import android.os.Build;
+import android.util.Log;
+import android.webkit.WebResourceError;
+import android.webkit.WebResourceRequest;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+import io.flutter.plugin.common.MethodChannel;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FlutterWebViewClient extends WebViewClient {
+  private static final String TAG = FlutterWebViewClient.class.getSimpleName();
+
+  private final MethodChannel methodChannel;
+
+  private Boolean shouldOverrideUrlLoading = false;
+
+  FlutterWebViewClient(MethodChannel methodChannel) {
+    assert methodChannel != null;
+    this.methodChannel = methodChannel;
+  }
+
+  @Override
+  public boolean shouldOverrideUrlLoading(WebView view, String url) {
+    return onShouldOverrideUrlLoading(view, url);
+  }
+
+  @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+  @Override
+  public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
+    return onShouldOverrideUrlLoading(view, request.getUrl().toString());
+  }
+
+  private boolean onShouldOverrideUrlLoading(final WebView view, final String url) {
+    if (shouldOverrideUrlLoading) {
+      Map<String, Object> args = new HashMap<>();
+      args.put("url", url);
+      methodChannel.invokeMethod(
```

Comment:
It is unfortunately very awkward to use an asynchronous method call here when the enclosing method is synchronous. For times like this I wish there were a way to make methodChannel calls synchronous. My application needs to override UrlLoading as well as intercept requests and provide alternate resources. However, so far it does not feel very feasible to implement with the current method channel system. I wonder if there plans or proposals to work around this. I suspect this specific class of problem will make several other integrations difficult.